### PR TITLE
Changed the router to default to roundrobin when weights are used

### PR DESCRIPTION
**Pull request by knobunc**
If the route has weights set for the services it is associated with,
then we will set the default load balance policy to RoundRobin if no
policy is set with an annotation or as a global default with an
environment variable.  Without this change the user would need to both
set the weights, and then set an annotation to change the default
balancing algorithm... which people almost always forgot to do.

For bug 1416869 (https://bugzilla.redhat.com/show_bug.cgi?id=1416869)

### CODE REVIEW

**Comment by knobunc** [comment]
@openshift/networking PTAL

**Comment by knobunc** [comment]
[test]

File: pkg/router/template/router.go [hunk 1]
```diff
@@ -841,23 +844,26 @@ func generateDestCertKey(config *ServiceAliasConfig) string {
...
-func getServiceUnits(route *routeapi.Route) map[string]int32 {
+func getServiceUnits(route *routeapi.Route) (map[string]int32, bool) {
 	serviceUnits := make(map[string]int32)
+	hasWeights := false
 	key := fmt.Sprintf("%s/%s", route.Namespace, route.Spec.To.Name)
 	if route.Spec.To.Weight == nil {
 		serviceUnits[key] = 0
 	} else {
 		serviceUnits[key] = *route.Spec.To.Weight
+		hasWeights = true
```
Comment (smarterclayton) [inline pkg/router/template/router.go]: Shouldn't this check be that there is more than one service?  A route with weight 100 and one backend should behave exactly the same as a route with weight 1 and one backend.

**Comment by smarterclayton** [comment]
change the e2e test for weighted router to remove the balance annotations

**Comment by knobunc** [comment]
@smarterclayton Good call.  I was contemplating the tests this morning, and that's a clean way to do it.

**Comment by smarterclayton** [comment]
In retrospect we probably could have caught that the review.  It did not
jump out at me at the time.

On Feb 2, 2017, at 9:46 AM, OpenShift Bot <notifications@github.com> wrote:

continuous-integration/openshift-jenkins/test Running (
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/13507/) (Base
Commit: cfb50d9
<https://github.com/openshift/origin/commit/cfb50d9156c6f5991a868a8d272e6b5503bebbae>
)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/12752#issuecomment-276976540>,
or mute
the thread

**Comment by knobunc** [comment]
[test] last failure was due to subscription-manager not getting a subscription

**Comment by knobunc** [comment]
[test] last was (now resolved) flake #12736

**Comment by knobunc** [comment]
[merge]

**Comment by smarterclayton** [comment]
Saw this show up https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_gce/278/testReport/junit/(root)/Extended/_networking__router__weighted_openshift_router_The_HAProxy_router_should_appropriately_serve_a_route_that_points_to_two_services/

Have never seen it before this PR merged.

**Pull request merged**